File: playwright/TC-func-REQ-SAF-004-17.ts
import { test, expect } from '@playwright/test';

test("TC-func-REQ-SAF-004-17 - Short circuit fault detection and recovery", async ({ page }) => {
  // Precondition: Vehicle system powered on and in normal operation, Fuel sensor connected and providing valid voltage 4.0V, Vehicle speed set to 25 km/h
  // Assume preconditions are already met.

  // Step 1: Set sensor voltage to 0.3V to simulate short circuit fault
  await page.locator('[data-testid="sensor-voltage-input"]').fill('0.3');
  await page.locator('[data-testid="sensor-voltage-submit"]').click();
  // Expect system detects voltage below 0.4V and activates fault detection
  const faultDetection = page.locator('[data-testid="fault-detection-status"]');
  await expect(faultDetection).toContainText('Fault detected');

  // Step 2: Verify fault code FUEL_SENSOR_FAILURE is stored
  const faultCodeList = page.locator('[data-testid="fault-code-list"]');
  await expect(faultCodeList).toContainText('FUEL_SENSOR_FAILURE');

  // Step 3: Verify continuous low-fuel warning is displayed
  const lowFuelWarning = page.locator('[data-testid="low-fuel-warning"]');
  await expect(lowFuelWarning).toBeVisible();

  // Step 4: Verify diagnostic message FUEL SENSOR ERROR - SERVICE REQUIRED is displayed
  const diagnosticMessage = page.locator('[data-testid="diagnostic-message"]');
  await expect(diagnosticMessage).toContainText('FUEL SENSOR ERROR - SERVICE REQUIRED');

  // Step 5: Verify fault log entry contains timestamp, fault code, and vehicle state
  const faultLog = page.locator('[data-testid="fault-log"]');
  await expect(faultLog).toContainText('FUEL_SENSOR_FAILURE');
  await expect(faultLog).toContainText('Timestamp');
  await expect(faultLog).toContainText('Vehicle state');

  // Step 6: Set sensor voltage back to 4.0V within 5 seconds and provide 3 consecutive valid readings
  await page.locator('[data-testid="sensor-voltage-input"]').fill('4.0');
  await page.locator('[data-testid="sensor-voltage-submit"]').click();
  await page.waitForTimeout(5000);
  for (let i = 0; i < 3; i++) {
    await page.locator('[data-testid="sensor-voltage-input"]').fill('4.0');
    await page.locator('[data-testid="sensor-voltage-submit"]').click();
    await page.waitForTimeout(1000);
  }
  // Expect system clears fault
  await expect(faultDetection).toContainText('No fault');

  // Step 7: Verify fault code cleared
  await expect(faultCodeList).not.toContainText('FUEL_SENSOR_FAILURE');

  // Step 8: Verify low-fuel warning inactive
  await expect(lowFuelWarning).not.toBeVisible();

  // Step 9: Verify diagnostic message cleared
  await expect(diagnosticMessage).not.toBeVisible();
});

File: selenium/TC-func-REQ-SAF-004-17.java
import org.openqa.selenium.*;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.support.ui.*;
import org.testng.Assert;
import org.testng.annotations.*;
import java.time.Duration;

public class TC-func-REQ-SAF-004-17_Test {
    private WebDriver driver;
    private WebDriverWait wait;

    @BeforeMethod
    public void setUp() {
        driver = new ChromeDriver();
        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        driver.get("http://vehicle-system-url"); // placeholder URL
    }

    @AfterMethod
    public void tearDown() {
        if (driver != null) {
            driver.quit();
        }
    }

    @Test
    public void test_TC-func-REQ-SAF-004-17() {
        // Step 1: Set sensor voltage to 0.3V to simulate short circuit fault
        WebElement voltageInput = wait.until(ExpectedConditions.elementToBeClickable(By.cssSelector("[data-testid='TODO']")));
        voltageInput.clear();
        voltageInput.sendKeys("0.3");
        // Wait for system to detect voltage below 0.4V
        wait.until(ExpectedConditions.textToBePresentInElementLocated(By.cssSelector("[data-testid='TODO']"), "Voltage below 0.4V"));

        // Step 2: Verify fault code FUEL_SENSOR_FAILURE is stored
        WebElement faultCodeElement = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[data-testid='TODO']")));
        Assert.assertTrue(faultCodeElement.getText().contains("FUEL_SENSOR_FAILURE"), "Fault code FUEL_SENSOR_FAILURE not found");

        // Step 3: Verify continuous low-fuel warning is displayed
        WebElement warningElement = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[data-testid='TODO']")));
        Assert.assertTrue(warningElement.getText().contains("Low-fuel warning"), "Low-fuel warning not active");

        // Step 4: Verify diagnostic message FUEL SENSOR ERROR - SERVICE REQUIRED is displayed
        WebElement diagMessageElement = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[data-testid='TODO']")));
        Assert.assertTrue(diagMessageElement.getText().contains("FUEL SENSOR ERROR - SERVICE REQUIRED"), "Diagnostic message not displayed");

        // Step 5: Verify fault log entry contains timestamp, fault code, and vehicle state
        WebElement logEntryElement = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[data-testid='TODO']")));
        String logText = logEntryElement.getText();
        Assert.assertTrue(logText.matches(".*\\d{4}-\\d{2}-\\d{2}.*"), "Timestamp missing in log");
        Assert.assertTrue(logText.contains("FUEL_SENSOR_FAILURE"), "Fault code missing in log");
        Assert.assertTrue(logText.contains("Vehicle state"), "Vehicle state missing in log");

        // Step 6: Set sensor voltage back to 4.0V within 5 seconds and provide 3 consecutive valid readings
        WebElement voltageInput2 = wait.until(ExpectedConditions.elementToBeClickable(By.cssSelector("[data-testid='TODO']")));
        voltageInput2.clear();
        voltageInput2.sendKeys("4.0");
        try { Thread.sleep(5000); } catch (InterruptedException e) {}
        for (int i = 0; i < 3; i++) {
            voltageInput2.clear();
            voltageInput2.sendKeys("4.0");
            try { Thread.sleep(1000); } catch (InterruptedException e) {}
        }

        // Step 7: Verify fault code cleared
        wait.until(ExpectedConditions.invisibilityOfElementLocated(By.cssSelector("[data-testid='TODO']")));
        WebElement faultCodeElementAfter = driver.findElement(By.cssSelector("[data-testid='TODO']"));
        Assert.assertFalse(faultCodeElementAfter.getText().contains("FUEL_SENSOR_FAILURE"), "Fault code not cleared");

        // Step 8: Verify low-fuel warning inactive
        WebElement warningElementAfter = driver.findElement(By.cssSelector("[data-testid='TODO']"));
        Assert.assertFalse(warningElementAfter.getText().contains("Low-fuel warning"), "Low-fuel warning still active");

        // Step 9: Verify diagnostic message cleared
        WebElement diagMessageElementAfter = driver.findElement(By.cssSelector("[data-testid='TODO']"));
        Assert.assertFalse(diagMessageElementAfter.isDisplayed(), "Diagnostic message not cleared");
    }
}